### PR TITLE
improve eventsource tests - part 1

### PR DESCRIPTION
**Pull request by TimothyMothra**
#3561.

## Changes
- fixed `OpenTelemetrySdkEventSource.ExporterErrorResult` by removing ToString.
  This method must call WriteEvent with the same types in the method signature.
- removed if statement from `EventSourceTestHelper`, 
   previously `OpenTelemetrySdkEventSource.ExporterErrorResult` was being excluded from tests.
- add missing test for `PrometheusExporterEventSource`.
- use `EventSource.SetCurrentThreadActivityId` to set a unique ActivityId before each EventSource invoke.
  This allows us to search the listener by the ActivityId. 
  This is important because EventSource will write an error if the log fails and we need that error message to fix issues.

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

## Todo
- after these changes are accepted, need to make the same changes in the [Contrib repo ](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/test/OpenTelemetry.Contrib.Tests.Shared/EventSourceTestHelper.cs)

### CODE REVIEW

File: test/OpenTelemetry.Exporter.Prometheus.HttpListener.Tests/EventSourceTest.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// <copyright file="EventSourceTest.cs" company="OpenTelemetry Authors">
```
Comment (alanwest) [inline test/OpenTelemetry.Exporter.Prometheus.HttpListener.Tests/EventSourceTest.cs:1]: Makes sense to move this file to the new common test project so it runs both `net462` and `net6.0` targets.

**Review by alanwest — APPROVED** [review]
LGTM. Just a small comment now that there's an `OpenTelemetry.Exporter.Prometheus.Tests` project

**Comment by cijothomas** [comment]
Failed OpenTelemetry.Tests.EventSourceTest.EventSourceTest_OpenTelemetrySdkEventSource [59 ms]
  Error Message:
   System.Exception : Method 'OpenTelemetrySdkEventSource.ShutdownEvent' is implemented incorrectly.
---- System.Exception : VerifyEventId Failed: expected: '2' actual: '40'
  Stack Trace:
     at OpenTelemetry.Tests.EventSourceTestHelper.VerifyMethodImplementation(EventSource eventSource, MethodInfo eventMethod) in D:\a\opentelemetry-dotnet\opentelemetry-dotnet\test\OpenTelemetry.Tests\Shared\EventSourceTestHelper.cs:line 68
   at OpenTelemetry.Tests.EventSourceTestHelper.MethodsAreImplementedConsistentlyWithTheirAttributes(EventSource eventSource) in D:\a\opentelemetry-dotnet\opentelemetry-dotnet\test\OpenTelemetry.Tests\Shared\EventSourceTestHelper.cs:line 32
   at OpenTelemetry.Tests.EventSourceTest.EventSourceTest_OpenTelemetrySdkEventSource() in D:\a\opentelemetry-dotnet\opentelemetry-dotnet\test\OpenTelemetry.Tests\EventSourceTest.cs:line 40
----- Inner Stack Trace -----
   at OpenTelemetry.Tests.EventSourceTestHelper.AssertEqual[T](String methodName, T expected, T actual) in D:\a\opentelemetry-dotnet\opentelemetry-dotnet\test\OpenTelemetry.Tests\Shared\EventSourceTestHelper.cs:line 134
   at OpenTelemetry.Tests.EventSourceTestHelper.VerifyEventId(MethodInfo eventMethod, EventWrittenEventArgs actualEvent) in D:\a\opentelemetry-dotnet\opentelemetry-dotnet\test\OpenTelemetry.Tests\Shared\EventSourceTestHelper.cs:line 106
   at OpenTelemetry.Tests.EventSourceTestHelper.VerifyMethodImplementation(EventSource eventSource, MethodInfo eventMethod) in D:\a\opentelemetry-dotnet\opentelemetry-dotnet\test\OpenTelemetry.Tests\Shared\EventSourceTestHelper.cs:line 60

Failed!  - Failed:     1, Passed:   792, Skipped:     3, Total:   796, Duration: 1 m 43 s - OpenTelemetry.Tests.dll (net6.0)
@TimothyMothra

**Comment by reyang** [comment]
What's the overall direction regarding internal troubleshooting logs?
If the answer is "it's clear that we should move to ILogger" (#1258), would this PR be a throwaway work?
If the answer is "it's clear that we will stick with EventSource", could someone reply to #1258 and resolve it?
If the answer is "it's unclear, we don't know what's the direction", does it make sense to clarify the direction first, or just add some tests right now - without knowing whether these are throwaway work or not?

**Comment by TimothyMothra** [comment]
> What's the overall direction regarding internal troubleshooting logs? If the answer is "it's clear that we should move to ILogger" (#1258), would this PR be a throwaway work? If the answer is "it's clear that we will stick with EventSource", could someone reply to #1258 and resolve it? If the answer is "it's unclear, we don't know what's the direction", does it make sense to clarify the direction first, or just add some tests right now - without knowing whether these are throwaway work or not?

I read the thread on #1258 and I didn't see any consensus to switch to ILogger. I believe that there is no change, and the current direction is to use EventSource. I think this idea has merit, but would need more of a design spec and/or POC to continue the conversation.

My PR here assumes there is no change to internal logging. I'm fixing a bug in one EventSource class, modifying the eventsourcetest helper to better catch bugs, and adding a missing eventsource test class. :)

**Comment by TimothyMothra** [comment]
> Failed OpenTelemetry.Tests.EventSourceTest.EventSourceTest_OpenTelemetrySdkEventSource [59 ms] Error Message: System.Exception : Method 'OpenTelemetrySdkEventSource.ShutdownEvent' is implemented incorrectly. ---- System.Exception : VerifyEventId Failed: expected: '2' actual: '40' 

This occurred in the CodeCov run but not the other test runs, which suggests it's in intermittent bug probably due to concurrent tests.
I'm not sure how this could happen because I expected setting an ActivityId on the current test's thread would make this situation impossible.

I can make more changes to the TestHelper to make this less likely.

File: src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs [hunk 2]
```diff
@@ -195,7 +195,7 @@ public void ShutdownEvent(int spansLeftUnprocessed)
         [Event(3, Message = "Exporter returned error '{0}'.", Level = EventLevel.Warning)]
         public void ExporterErrorResult(ExportResult exportResult)
         {
-            this.WriteEvent(3, exportResult.ToString());
+            this.WriteEvent(3, exportResult);
```
Comment (alanwest) [inline src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:198]: Does this fix the behavior we're seeing here https://github.com/open-telemetry/opentelemetry-dotnet/issues/3619#issuecomment-1234724122? If so, we should get this landed.

Comment (cijothomas) [reply src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:198]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/598 and old PR doing similar stuff.

Comment (TimothyMothra) [reply src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:198]: @alanwest, I don't have a straightforward repo to confirm if this is the issue seen in #3619. 

@cijothomas, I verified that my change works, and can confirm that the `ExportResult` enum is successfully converted to string.
I don't think the `[NonEvent]` pattern in #598 is necessary. But I'm happy to change if that's the preferred pattern. :)

Comment (cijothomas) [reply src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:198]: nonevent pattern is to save string alloc if none listening.

Does this PR's change create the string if none is listening?

Comment (TimothyMothra) [reply src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:207]: changed to `NonEvent` pattern

**Pull request merged**